### PR TITLE
5 fix esphome bt proxy compatibility

### PR DESCRIPTION
**Pull request by filmkorn**
Requires: 
- Home assistant 2024.10 (manifest `requirements` syntax )
- https://github.com/filmkorn/pyanova-nano/pull/14

### CODE REVIEW

File: custom_components/anova_nano/manifest.json [hunk 1]
```diff
@@ -13,6 +13,6 @@
   "documentation": "https://github.com/mcolyer/hacs-anova-nano",
   "iot_class": "local_polling",
   "issue_tracker": "https://github.com/mcolyer/hacs-anova-nano/issues",
-  "requirements": ["pyanova-nano==0.2.1"],
+  "requirements": ["pyanova-nano @ git+https://github.com/filmkorn/pyanova-nano.git@13_esphome_backend_compatibility"],
```
Comment (filmkorn) [inline custom_components/anova_nano/manifest.json:16]: Bump to released version.

File: custom_components/anova_nano/manifest.json [hunk 2]
```diff
@@ -13,6 +13,6 @@
...
   "version": "0.1.0"
```
Comment (filmkorn) [inline custom_components/anova_nano/manifest.json:17]: Bump version

**Comment by filmkorn** [comment]
@mcolyer When you have a chance, could you try this out? Must say I'm not sure if the `requirements` syntax works - it seems to have changed with the migration to uv. Alternatively I was able to update pyanova-nano with:
```
uv pip install git+https://github.com/filmkorn/pyanova-nano.git@13_esphome_backend_compatibility --force-reinstall --no-deps
```

**Comment by filmkorn** [comment]
>  separately have a pr going to fix the version numbers and switch this repo to uv.

Thanks for that! And apologies for neglecting the tests. 

TBH I've not yet figured out how to use the dev container. So I've been restarting my "production" home assistant a million times recently.

**Comment by mcolyer** [comment]
Okay I merged in my fixes if you want to rebase.

**Comment by mcolyer** [comment]
So I don't see any errors in the log and I can add the device and it doesn't activate.

**Comment by filmkorn** [comment]
Thanks for trying. Can you try enabling debug logging on the integration?

image

**Comment by filmkorn** [comment]
I'm convinced hass or uv doesn't install the correct requirement using the `pyanova-nano @ git+https://github.com/filmkorn/pyanova-nano.git@13_esphome_backend_compatibility` syntax. I've released the change in `pyanova-nano` - I hope that this finally works with your ESPHome BT Proxy as well. If not - the debug logs are now pretty verbose and should give us some clues as to what might be going wrong.

**Comment by mcolyer** [comment]
Okay got the debug logs loaded and have this:

```
2024-10-04 20:01:20.815 DEBUG (MainThread) [custom_components.anova_nano] PyAnova - C4:64:E3:13:57:9E: Failed to connect after 9 attempt(s): Error ESP_GATT_CONN_FAIL_ESTABLISH while connecting: Connection failed to establish: Interference/range; External Bluetooth adapter w/extension may help; Extension cables reduce USB 3 port interference
Traceback (most recent call last):
File "/usr/local/lib/python3.12/site-packages/bleak_retry_connector/__init__.py", line 363, in establish_connection
await client.connect(
File "/usr/local/lib/python3.12/site-packages/habluetooth/wrappers.py", line 311, in connect
connected = await super().connect(**kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/bleak/__init__.py", line 615, in connect
return await self._backend.connect(**kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/bleak_esphome/backend/client.py", line 77, in _async_wrap_bluetooth_operation
return await func(self, *args, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/bleak_esphome/backend/client.py", line 327, in connect
await connected_future
bleak.exc.BleakError: Error ESP_GATT_CONN_FAIL_ESTABLISH while connecting: Connection failed to establish
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
File "/config/custom_components/anova_nano/coordinator.py", line 83, in _connect
await self._client.connect(
File "/usr/local/lib/python3.12/site-packages/pyanova_nano/client.py", line 147, in connect
await self._connect(self._device, timeout_seconds=timeout_seconds)
File "/usr/local/lib/python3.12/site-packages/pyanova_nano/client.py", line 167, in _connect
self._client = await establish_connection(
^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/bleak_retry_connector/__init__.py", line 472, in establish_connection
_raise_if_needed(name, device.address, exc)
File "/usr/local/lib/python3.12/site-packages/bleak_retry_connector/__init__.py", line 336, in _raise_if_needed
raise BleakAbortedError(f"{msg}: {ABORT_ADVICE}") from exc
bleak_retry_connector.BleakAbortedError: PyAnova - C4:64:E3:13:57:9E: Failed to connect after 9 attempt(s): Error ESP_GATT_CONN_FAIL_ESTABLISH while connecting: Connection failed to establish: Interference/range; External Bluetooth adapter w/extension may help; Extension cables reduce USB 3 port interference
```

**Comment by mcolyer** [comment]
It works! So it was too far before. However one odd thing I noticed:

image

My target temp is 150 degrees F on the device but it shows as 302 F on the dashboard. I think it's mistakenly converting 150 into 302 F. Also I think there's a similar problem with the water temperature sensor, as it's currently out of the water and would guess it's supposed to be 80 and not 174.

**Comment by mcolyer** [comment]
It also seems to only reconnect after I unplug the anova nano if I reload the custom integration, is that expected?

**Comment by filmkorn** [comment]
> My target temp is 150 degrees F on the device but it shows as 302 F on the dashboard.

Oh that's certainly a unit conversion issue. I suppose I forgot to report the native unit on the AnovaNanoNumberEntity as it's done on the AnovaNanoSensor: 

https://github.com/mcolyer/hacs-anova-nano/blob/4e82f5f7ea07e85de609b242063b173484b0e9a0/custom_components/anova_nano/sensor.py#L112-L116

> It also seems to only reconnect after I unplug the anova nano if I reload the custom integration, is that expected?

No that's not expected. I also noticed this and I wanted to create an issue for it. There could be two issues:
- Client (Bleak) is not disconnected cleanly - eg when the Adapter or Bluetooth Proxy disconnects. In theory the `on_disconnect` callback should be called that tells hass to rediscover the device but something isn't working.

https://github.com/mcolyer/hacs-anova-nano/blob/4e82f5f7ea07e85de609b242063b173484b0e9a0/custom_components/anova_nano/coordinator.py#L95-L108

- The unclean disconnection leaves the Anova Nano in a state where it doesn't seem to advertise itself anymore. I'm not sure how we'd fix this.

**Pull request merged**